### PR TITLE
PIM-7084: add clean command for deleted attributes

### PR DESCRIPTION
**Pull request by juliensnz**
## TL;DR
- The solution to fix a memory leak can be "I will not fix this memory leak"
- To avoid having problems with memory, split your big task into smaller ones
- The solution can come from crazy experiments

## History of this PR

We have a simple mission for this SLA: iterate over all products of the catalog and save them. Nothing fancy, just load and save a big quantity of products. 

We started by doing it with classical:
- filter products with the PQB
- iterate
- save them with product and product model saver

Everything was fine until we looked at the memory consumption: the command was leaking about 2mb every 10 products.

With this kind of numbers, we already know that the command will crash really soon for big clients and we need to take care of it before it comes back to support.

## Trying to fix the memory leak

The most obvious solution was to try to fix this memory leak so we took [php-meminfo](https://github.com/BitOne/php-meminfo) out and found multiple problems:

- [UploadChecker](https://github.com/akeneo/pim-enterprise-dev/blob/cfa4a611709766454402981cf64aa2dd69d9bfb2/src/PimEnterprise/Component/ProductAsset/Upload/UploadChecker.php#L43) was keeping some locales in memory for nothing
- the old [UserPreferencesSubscriber](https://github.com/akeneo/pim-community-dev/blob/234b4d49f08cdb4b7edb97e39342393da92bc19e/src/Pim/Bundle/UserBundle/EventSubscriber/UserPreferencesSubscriber.php#L74) was keeping references to uow for nothing
- etc

With all these fixes, the command was still leaking and we were not able to clear the doctrine cache (with the CacheClearer) because we had unmanaged entities in the uow (famous `A new entity was found through the relationship...`).

We were a bit stuck at this point.

## Seeking for (crazy) alternatives

After all those efforts, we were a bit desperate and clueless about how to fix this problem.

So we took a look at some unconventional solutions:

- Create a command that launch the 'pim:product:updates' without any update -> it was working but reaaally slow (1s per products) and was not supporting product models
- Create a command to launch a mass edit with no update -> it was efficient but the mass edit skips when there is no updates so not usable in place
- Create a command to call the API and ask for empty updates -> the API cannot work unothenticated
- Create a dedicated job with an empty processor -> This solution could have worked but we didn't wanted to create a job to fix a temporary bug

## The final solution

With all those experimentations we saw that the key was to split the job into smaller tasks:

- We created a command that get all product[model]s ids (using a cursor) and group them by `batch_size`
- A second command that takes a batch of product[model]s ids and does the job (saveAll)

That way, the second command leaks but we don't care as it's only on a very small batch of items (10 in this case). The first command is in charge of the interactions with the user (progress bar, confirmation, etc) and does not leak (we don't load anything). The batch size is configurable.

## Conclusion

This solution is really close to something we discussed with our beloved CTO lately: stop doing big job in one run to avoid having to manage memory leaks anymore. Finding a nice solution to split big jobs into smaller one seems to be a nice solution and easy to implement.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | NA
| Added legacy Behats               | NA
| Added acceptance tests            | NA
| Added integration tests           | NA
| Changelog updated                 | Yes
| Review and 2 GTM                  | Yes
| Micro Demo to the PO (Story only) | Yes
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Command/CleanProductValueDuplicationsCommand.php [hunk 1]
```diff
@@ -1,177 +0,0 @@
-<?php
```
Comment (juliensnz) [inline src/Pim/Bundle/CatalogBundle/Command/CleanProductValueDuplicationsCommand.php]: This command was made to clean the product value table that doesn't exists anymore

File: src/Pim/Bundle/CatalogBundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            ->setDescription('Removes all values of deleted attributes on all products and product models');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output): void
+    {
+        $productBatchSize = $this->getContainer()->hasParameter('pim_job_product_batch_size') ?
+            $this->getContainer()->getParameter('pim_job_product_batch_size') :
```
Comment (juliensnz) [inline src/Pim/Bundle/CatalogBundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php]: the parameter `pim_job_product_batch_size` is still not present in the 2.0 branch. With this line, we define a fallback.

File: src/Pim/Bundle/UserBundle/EventSubscriber/UserPreferencesSubscriber.php [hunk 3]
```diff
@@ -70,13 +64,13 @@ public function getSubscribedEvents()
      */
     public function onFlush(OnFlushEventArgs $args)
     {
-        $this->manager = $args->getEntityManager();
```
Comment (juliensnz) [inline src/Pim/Bundle/UserBundle/EventSubscriber/UserPreferencesSubscriber.php]: we were setting the uow as instance variable outside a constructor. The uow was linked to this service and was keeping elements attached to him.

File: src/Pim/Bundle/CatalogBundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        $this
+            ->setName('pim:product:clean-removed-attributes')
+            ->setDescription('Removes all values of deleted attributes on all products and product models');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output): void
+    {
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php]: Maybe split the body of this function in multiple private for the sake of clarity.

File: src/Pim/Bundle/CatalogBundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        , false);
+
+        if (!$answer) {
+            $io->text('That\'s ok, see you!');
+
+            return;
+        }
+
+        $io->text([
+            'Ok, let\'s go!',
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php]: 👍

**Comment by oliverde8** [comment]
Hi, 

I have been encountering memory issues & performance issues with Akeneo since it came out. What I read is : 
> we were not able to clear the doctrine cache

Which is a point I have raised on a few occasions when we met. Doctrine is fine when you do mass edits in a single table without having foreign resources being loaded. You can with a bit of effort manage the memory usage. But Doctrine don't work well for complex mass actions as you can't detach objects in controlled way. 

One of the points of using Doctrine is to have generic support for mysql & other sql databases. Akeneo doesn't do that anyway. 

I believe there could be important performance gains if some of these mass actions (such as imports) could go around Doctrine. We are importing 500K products per hour in other tools in mysql (with pre validation in php) while Akeneo can barely do do 50K per hour with the same products. 

My thought on the subject, because splitting files, we do that already by obligation.

**Comment by juliensnz** [comment]
Hello @oliverde8,

You are totally right, doctrine is not made for big batches of complex entities. We try to fix this situation. But as you can see, it's not really easy on the Product entity. A lot of things are linked to doctrine here and we cannot get rid of it right away. This is clearly a first step and we don't think that it's a final one.

BTW, your performance comparison is not really fair as we don't only do validation with the hydrated objects. We do a lot of computation: create proposals, compute completeness, we apply rules, we apply rights, etc. Doing mysql insert is cheap and easy but not really extensible :)

I hope that one day, we will be able to get rid of doctrine ORM on this kind of operations but until then, we will have to live with it :)

**Comment by oliverde8** [comment]
Hi, 

I am glad to hear it is something you are at least thinking of, as last time I had talked about it it wasn't something you were considering. And indeed doctrine is great to start with as it allows to get something going very fast. 

Well performance comparison was without rules in akeneo and in my example actually the validations goes farther then what akeneo was doing. Still I think aiming at importing a few hundred K products(or models) per hour with a simple catalog keeping existing functionality is doable. 

Thanks for the response, have a nice day 😄

File: src/Pim/Bundle/CatalogBundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        $progressBar->start();
+
+        $productToCleanCount = 0;
+
+        $rootDir = $this->getContainer()->get('kernel')->getRootDir();
+        foreach ($products as $product) {
+            $productIds[] = IdEncoder::encode($product instanceof ProductModel ? 'product_model' : 'product', $product->getId());
+            $productToCleanCount++;
+            if (0 === $productToCleanCount % $productBatchSize) {
+                $process = new Process([sprintf('%s/../bin/console', $rootDir), 'pim:product:refresh', '--env=prod', implode(',', $productIds)]);
```
[the remark this answers is not in the document]
Comment (juliensnz) [reply src/Pim/Bundle/CatalogBundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php]: yeah, good idea

File: src/Pim/Bundle/CatalogBundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+    /**
+     * Lanches the clean command on given ids
+     *
+     * @param array  $productIds
+     * @param string $env
+     */
+    private function launchCleanTask(array $productIds, string $env)
+    {
+        $rootDir = $this->getContainer()->get('kernel')->getRootDir();
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Command/CleanRemovedAttributesFromProductAndProductModelCommand.php]: In the other privates you chose to not call the container and pass the dependencies as arguments. It would make sense here too.

File: src/Pim/Component/Connector/Writer/Database/ProductWriter.php [hunk 8]
```diff
@@ -48,7 +42,6 @@ public function __construct(
     ) {
         $this->versionManager = $versionManager;
         $this->productSaver = $productSaver;
-        $this->cacheClearer = $cacheClearer;
```
Comment (phaseinducer) [inline src/Pim/Component/Connector/Writer/Database/ProductWriter.php]: It's a BC break.

Comment (juliensnz) [reply src/Pim/Component/Connector/Writer/Database/ProductWriter.php]: no as the constructor parameter stay the same
